xwiki-platform-core/xwiki-platform-localization/xwiki-platform-localization-rest/src/main/java/org/xwiki/localization/internal/DefaultLocalizationSource.java

```diff
@@ -0,0 +1,115 @@
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+import javax.ws.rs.core.Response;
+
+import org.apache.commons.lang3.LocaleUtils;
+import org.slf4j.Logger;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.localization.LocalizationContext;
+import org.xwiki.localization.LocalizationManager;
+import org.xwiki.localization.Translation;
+import org.xwiki.localization.rest.LocalizationSource;
+import org.xwiki.rest.XWikiRestComponent;
+
+import com.fasterxml.jackson.databind.node.JsonNodeFactory;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.xpn.xwiki.XWikiContext;
+
+/**
+ * Default implementation of {@link LocalizationSource}.
+ *
+ * @version $Id$
+ * @since 13.3RC1
+ */
+@Component
+@Named("org.xwiki.localization.internal.DefaultLocalizationSource")
+@Singleton
+public class DefaultLocalizationSource implements LocalizationSource, XWikiRestComponent
+{
+    @Inject
+    private LocalizationContext localizationContext;
+
+    @Inject
+    private LocalizationManager localizationManager;
+
+    @Inject
+    private Provider<XWikiContext> contextProvider;
+
+    @Inject
+    private Logger logger;
+
+    @Override
+    public Response translations(String wikiId, String localeString, String prefix, List<String> keys)
+    {
+        XWikiContext xWikiContext = this.contextProvider.get();
+
+        // Save the current context wiki id. It will be restored on the method is finished.
+        String oldWikiId = xWikiContext.getWikiId();
+        try {
+            // Set the requested wiki id.
+            xWikiContext.setWikiId(wikiId);
+
+            // Resolve the local.
+            Locale locale;
+            if (localeString != null) {
+                locale = LocaleUtils.toLocale(localeString);
+            } else {
+                locale = this.localizationContext.getCurrentLocale();
+            }
+
+            // Computes the prefix.
+            String cleanPrefix = Objects.toString(prefix, "");
+
+            ObjectNode result = JsonNodeFactory.instance.objectNode();
+
+            // Resolve the raw translation of the requested keys.
+            for (String key : keys) {
+                if (key != null) {
+                    String fullKey = cleanPrefix + key;
+                    Translation translation = this.localizationManager.getTranslation(fullKey, locale);
+                    if (translation != null) {
+                        Object rawSource = translation.getRawSource();
+                        result.put(fullKey, String.valueOf(rawSource));
+                    } else {
+                        // When a translation key is not found, it is still added to the result object, associated with
+                        // a null value.
+                        result.putNull(fullKey);
+                        this.logger.warn("Key [{}] not found for local [{}] in wiki [{}].", key, locale, wikiId);
```

Comment:
I would rather fall-back on the English translation. Actually, would be good to follow the full chain: locale, parent locale, .., root locale. Also, logging a warning for this feels to much because having translation keys with missing values for a specific locale is very common. @tmortagne WDYT?

Reply:
> would be good to follow the full chain: locale, parent locale, .., root locale

That's already what `LocalizationManager.getTranslation`. If you get null it means this tranlation simply does not exist at all (in this context at least).

Reply:
OK, good to know. I would still use debug level and not warning though.

Reply:
> I would still use debug level and not warning though.

Not sure, having a UI which request a non-existing translation seems like a mistake to me. Do you have a use case in mind ?

Reply:
Could happen with a cached (outdated) UI (JavaScript) and deleted translation keys that didn't went through deprecation. But indeed it would mean something is wrong (i.e. the front-end is not in sync with the back-end) so the warning could help the user raise and fix the problem.

I just don't want to see this warning when a translation is missing for a specific locale but it's available in English, but that's not the case AFAIU, so I'm fine with warning.

Reply:
I stayed with the warn for now.